compiler/fir/resolve/src/org/jetbrains/kotlin/fir/resolve/transformers/body/resolve/FirExpressionsResolveTransformer.kt

```diff
@@ -467,11 +467,21 @@ open class FirExpressionsResolveTransformer(transformer: FirBodyResolveTransform
         }
 
         return when {
-            assignIsResolvable && !lhsIsVar -> chooseAssign()
-            !assignIsResolvable && !operatorIsResolvable -> chooseAssign()
-            !assignIsResolvable && operatorIsResolvable -> chooseOperator()
-            assignIsResolvable && !operatorIsResolvable -> chooseAssign()
-            assignIsResolvable && operatorIsResolvable && !operatorReturnTypeMatches -> chooseAssign()
+            assignIsSuccessful && !lhsIsVar -> chooseAssign()
+            !assignIsSuccessful && !operatorIsSuccessful -> {
+                // If neither candidate is successful, choose whichever is resolved, prioritizing assign
+                val isAssignResolved = assignCallReference.safeAs<FirErrorReferenceWithCandidate>()?.diagnostic !is ConeUnresolvedNameError
+                val isOperatorResolved =
+                    operatorCallReference.safeAs<FirErrorReferenceWithCandidate>()?.diagnostic !is ConeUnresolvedNameError
+                if (isAssignResolved || !isOperatorResolved) {
+                    chooseAssign()
+                } else {
+                    chooseOperator()
+                }
+            }
```

Comment:
nit: I find it easier to follow with

```
when {
  isAssignResolved -> chooseAssign()
  isOperatorResolved -> chooseOperator()
  else -> chooseAssign()
}
```

Reply:
Thanks for the suggestion, I like that better too! I should've followed what you did instead of trying to be fancy.